numba/tests/test_unicode.py

```diff
@@ -338,6 +386,110 @@ def test_find(self, flags=no_pyobj_flags):
+                             error_msg.format(s, sub, py_result, c_result))
+
+    def test_count_with_start(self):
+        pyfunc = count_with_start_usecase
+        cfunc = njit(pyfunc)
+        error_msg = "%s\n%s" % ("'{0}'.py_count('{1}', {2}) = {3}",
+                                "'{0}'.c_count('{1}', {2}) = {4}")
+
+        for s, sub in UNICODE_COUNT_EXAMPLES:
+            for i in range(-18, 18):
+                py_result = pyfunc(s, sub, i)
+                c_result = cfunc(s, sub, i)
+                self.assertEqual(py_result, c_result,
+                                 error_msg.format(s, sub, i, py_result,
+                                                  c_result))
+
+            py_result = pyfunc(s, sub, None)
+            c_result = cfunc(s, sub, None)
+            self.assertEqual(py_result, c_result,
+                             error_msg.format(s, sub, None, py_result,
+                                              c_result))
+
+    def test_count_with_start_end(self):
+        pyfunc = count_with_start_end_usecase
+        cfunc = njit(pyfunc)
+        error_msg = "%s\n%s" % ("'{0}'.py_count('{1}', {2}, {3}) = {4}",
+                                "'{0}'.c_count('{1}', {2}, {3}) = {5}")
+
+        for s, sub in UNICODE_COUNT_EXAMPLES:
+            for i , j in product(range(-18, 18), (-18, 18)):
+                py_result = pyfunc(s, sub, i, j)
+                c_result = cfunc(s, sub, i, j)
+                self.assertEqual(py_result, c_result,
+                                 error_msg.format(s, sub, i, j, py_result,
+                                                  c_result))
+
+            for j in range(-18, 18):
+                py_result = pyfunc(s, sub, None, j)
+                c_result = cfunc(s, sub, None, j)
+                self.assertEqual(py_result, c_result,
+                                 error_msg.format(s, sub, None, j, py_result,
+                                                  c_result))
+
+            py_result = pyfunc(s, sub, None, None)
+            c_result = cfunc(s, sub, None, None)
+            self.assertEqual(py_result, c_result,
+                             error_msg.format(s, sub, None, None, py_result,
+                                              c_result))
+
+    def test_count_arg_type_check(self):
+        cfunc = njit(count_with_start_end_usecase)
+
+        with self.assertRaises(TypingError) as raises:
+            cfunc('ascii', 'c', 1, 0.5)
+        self.assertIn('The slice indices must be an Integer or None',
+                      str(raises.exception))
+
+        with self.assertRaises(TypingError) as raises:
+            cfunc('ascii', 'c', 1.2, 7)
+        self.assertIn('The slice indices must be an Integer or None',
+                      str(raises.exception))
+
+        with self.assertRaises(TypingError) as raises:
+            cfunc('ascii', 12, 1.2, 7)
+        self.assertIn('The slice indices must be an Integer or None',
```

Comment:
This isn't the first problem though is it? It is a problem, but the real issue is trying to `'ascii'.count(12)`?

Reply:
oh I see, this emulates python, slice raises ahead of type check on the counted string.

Reply:
It'd be nice if this:
```
cfunc('ascii', 12, 1, 7)
```
raised with an error message that was better than:
```
*** numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of BoundFunction((<class 'numba.types.misc.UnicodeType'>, 'count') for unicode_type) with parameters (int64, int64, int64)
```